proposals/0386-package-access-modifier.md

```diff
@@ -266,6 +263,88 @@ Potential solutions include introducing new keywords for specific access combina
+
+The above applies to a test target as well.  If a target contains package APIs that need to be tested, it would be recommended to split it into a "shell" that contains public APIs and a target that contains package APIs to be tested with the parameter `accessPublicOnly` set to the default value `false`.  This naturally eliminates the need to import the module with `@testable`.
+
+2. Sub-packages
+
+As a package becomes larger, it is inevitable for some targets to belong to a certain group within the package.  Even though it logically makes sense to move them out into a separate package, they often remain in the same package due to the maintenance and versioning overhead.  Currently each local package needs to be at the root directory regardless of the package dependencies, but we could allow a dependency package to reside in a sub-directory, which would make it easier for an existing package to convert into a structured top-level package with sub-packages.
+
+The top-level package and its 2 sub-packages would then be updated as follows.
+
+Top-level package in directory `gamePkg`:
+
+```
+let package = Package(
+    name: "gamePkg",
+    products: [
+        .library(name: "Game", targets: ["Game"]),
+    ],
+    dependencies: [
+        .package("A"),
+        .package("B")
+    ],
+    targets: [
+        .target(name: "Game",
+                dependencies: [
+                    .product(name: "CoreA", package: "A"),
+                    .product(name: "CoreB", package: "B")
+                ]),
+    ],
+)
+```
+
+Sub-package `A` in directory 'gamePkg/A':
+```
+let package = Package(
+        name: "A",
+        products: [
+            .library(name: "CoreA", targets: ["CoreA"]),
+        ],
+        targets: [
+            .target(name: "CoreA", dependencies: ["EngineA"]),
+            .target(name: "EngineA"),
+        ]
+      )
+```
+
+Sub-package `B` in directory `gamePkg/B`:
+
+```
+let package = Package(
+        name: "B",
+        products: [
+            .library(name: "CoreB", targets: ["CoreB"]),
+        ],
+        targets: [
+            .target(name: "CoreB", dependencies: ["EngineB"]),
+            .target(name: "EngineB"),
+        ]
+      ),
+
+```
+
+When building target `CoreA` in sub-package `A`, SwiftPM will pass `gamePkg.A` to `-package-name`; similarly, target `CoreB` will be built with `-package-name gamePkg.B`. The top level package ID is required to be unique (in case of SwiftPM, uniqueness is guaranteed via a registry), and appending a sub-package ID to it will ensure the uniqueness of each sub-package; if another top level package has the same ID as one of the sub-packages in this example, such as `A`, the `-package-name` input for the top level package `A` will be differentiated from the input `gamePkg.A` for the sub-package.
```

Comment:
Some reviewers were concerned about requiring manifests to be divided this way; I think the discussion here at least needs to address that argument.  Users seem to feel that splitting manifest files has substantial practical / maintenance costs.  See e.g. this comment: https://forums.swift.org/t/se-0386-package-access-modifier/62808/154.